man/man1/oci-network-config.1
.\" Process this file with
.\" groff -man -Tascii oci-network-config.1
.\"
.\" Copyright (c) 2017, 2022 Oracle and/or its affiliates. All rights reserved.
.\" Licensed under the Universal Permissive License v 1.0 as shown
.\" at http://oss.oracle.com/licenses/upl.
.\"
.TH OCI-NETWORK-CONFIG 1 "MAY 2018" Linux "User Manuals"
.SH NAME
oci-network-config \- configure network devices on Oracle Cloud Infrastructure compute instances
.SH SYNOPSIS

.B oci-network-config
usage

.B oci-network-config
show [-h|--help] [-I ITEM|--include ITEM] [-X ITEM|--exclude ITEM] [--output-mode MODE] [--details]

.B oci-network-config
show-os [-h|--help] [--output-mode MODE]

.B oci-network-config
show-vcns [-h|--help] [--output-mode MODE] [--details] [--ocid VCN_OCID] [--name VCN_NAME] [--no-truncate]

.B oci-network-config
show-subnets [-h|--help] [--output-mode MODE] [--details] [--ocid SUBNET_OCID] [--name SUBNET_NAME] [--no-truncate]

.B oci-network-config
show-vnics [-h|--help] [--output-mode MODE] [--details] [--ocid VNIC_OCID] [--name VNIC_NAME] [--ip-adress PRIMARY_IP] [--no-truncate]

.B oci-network-config
show-vnics-all [-h|--help] [--output-mode MODE]

.B oci-network-config
configure [-h|--help] [-n FORMAT|--namespace FORMAT] [-r |--start-sshd] [-I ITEM|--include ITEM] [-X ITEM|--exclude ITEM]

.B oci-network-config
unconfigure [-h|--help] [-I ITEM|--include ITEM] [-X ITEM|--exclude ITEM]

.B oci-network-config
attach-vnic [-h|--help] [-I IP_ADDR|--ip-addres IP_ADDR] [-i INDEX |--nic-index INDEX] [--subnet SUBNET]
[-n NAME |--name NAME] [--ipv4|-ipv4] [--ipv6|-ipv6] [--assign-public-ip] [--configure]

.B oci-network-config
detach-vnic [-h|--help] [-I IP_ADDR|--ip-addres IP_ADDR] [--ocid OCID]

.B oci-network-config
add-secondary-addr [-h|--help] [--ipv4|-ipv4] [--ipv6|-ipv6] [-I IP_ADDR|--ip-addres IP_ADDR] [--ocid OCID]

.B oci-network-config
remove-secondary-addr [-h|--help] [-I IP_ADDR|--ip-addres IP_ADDR]


.SH INSTALLATION and AUTHORIZATION
.B oci-network-config
requires the services of the python
.B oci-sdk
package.

The
.B oci-utils
package has a dependency on the
.B oci-sdk
package and the correct version of the python
.B oci-sdk
package is installed at
.B oci-utils
install time.

To use the Oracle Cloud Infrastructure services authorization is required. One option is direct authentication via a
config file.  See https://docs.us-phoenix-1.oraclecloud.com/Content/API/Concepts/sdkconfig.htm
for more information.
Another option for authorizing the use of Oracle Cloud Infrastructure services from an instance is
using Instance Principals.  The Oracle Cloud Infrastructure SDK does not need configuration on the
instance when the instance is part of a dynamic group that was granted access
to Oracle Cloud Infrastructure services.
See https://docs.us-phoenix-1.oraclecloud.com/Content/Identity/Tasks/callingservicesfrominstances.htm for further details.

.SH OCID

When the
.BR ocid(8)
service is enabled, it runs the network configuration utility each time the VNIC configuration metadata
of the instance changes.
It also restores the VNIC configuration after rebooting the system.  If
.BR ocid(8)
service is not enabled the
.B oci-network-config configure
 script must be run periodically to pick up changes in VNIC provisioning
(whether adding or deleting). Note that these IP configuration changes are not persistent,
the script must, at a minimum, be run after each startup.

To exclude VNICs, IP addresses or VLAN interfaces from automatic configuration use the -X (or --exclude) option.
Exclusions specified with this option are persistent so that they also apply when
.BR ocid(8)
runs this utility.  The exclusion remains in place until the VNIC/address/interface is re-included
using the -I (or --include) option.

.SH DESCRIPTION

The
.B oci-network-config
utility shows the current
Virtual interface Cards provisioned in the
Oracle Cloud Infrastructure
and configured on this instance. When a secondary VNIC is provisioned in Oracle Cloud Infrastructure it must be
explicitly configured on the instance using
.B oci-network-config
or similar commands.

By default this command displays the currently provisioned VNICs and the current IP configuration for this instance.
VNICs that are not yet configured are marked with
.B ADD
and IP configurations that no longer have an associated VNIC are marked with
.B DELETE.

Using the 
.B configure
sub-command, it configures VNICs that do not have an IP configuration and deletes the IP configurations of
VNICs that are not currently provisioned. This puts the instance IP configuration in sync with current
Oracle Cloud Infrastructure provisioning.

The configuring interfaces can optionally be placed inside separate network namespaces. This is necessary when
VNICs are in subnets (different VCNs) with overlapping address blocks and the network applications are not bound
directly to interfaces. Network namespaces require applications to be launched in them explicitly
(via 'ip netns exec <ns>') in order to establish association with the interface. When namespaces are not used,
policy-based routing is configured to provide a default route to the secondary VNIC\'s virtual router
(default gateway) when the VNIC\'s address is the source address.

Bare Metal secondary VNICs are configured using VLANs (where there is no corresponding physical interface).
These will look like 2 additional interfaces when showing IP links, with names like
.B MACVLAN_FORMAT
for the MAC VLAN and
.B VLAN_FORMAT
for the IP VLAN.

Using the
.B unconfigure
sub-command, it deletes all IP configuration for provisioned secondary VNICs except the ones excluded earlier.

The
.B attach-vnic
and
.B detach-vnic
sub-commands allows creating a new VNIC for the instance, detaching a VNIC from the instance,
or adding/removing private IP addresses to/from an existing VNIC.

The
.B add-secondary-addr
and
.B remove-secondary-addr
sub-commands allows adding/removing private IP addresses to/from an existing VNIC.


.TP
oci-network-config show
Show information on all provisioning and interface configuration. This is the default action if no options are given.

Columns:
    State     'ADD' indicates missing IP config, 'DELETE' missing VNIC, 'EXCL' excluded (-X)
    Link      interface (underlying physical)
    Status    state of interface
    Ip        IP address
    VNIC      VNIC friendly name
    MAC       MAC address
    Hostname  the configured hostname of that vNIC
    Subnet    subnet CIDR and name
    Router IP virutal router IP address
    Namespace namespace (if any)
    Index     interface index
    VLAN tag  VLAN tag
    VLAN      IP virtual LAN 
.RS
.IP "-h, --help"
Print a summary of the command line options.
.IP "-X ITEM, --exclude ITEM"
Persistently exclude
.B ITEM
from automatic configuration/deconfiguration. Use the --include option to include the
.B ITEM
again.
.B ITEM
can be a VNIC OCID, an IP address or a VLAN interface name.
.IP "-I ITEM, --include ITEM"
Include an
.B ITEM
that was previously excluded using the --exclude option in automatic configuration/deconfiguration.
.IP "--details"
Display detailed information.
.IP --output-mode={parsable,table,json,text}
Display list of block volumes as table, using json format, simple text or using parsable output
(default is table).
.RE
.TP
oci-network-config show-vcns
Shows vcn information

Columns:
    Name
    OCID
    ipv4 cidr block
    ivp6 cidr block
    ipv4 cidr blocks
    DNS label
    State
    Lifecycle state
.RS
.IP "-h, --help"
Print a summary of the command line options.
.IP "--ocid VCN_NAME"
Show only information of subnet matching name.
.IP "--ocid VCN_OCID"
Show  information of subnet matching ocid.
.IP "--details"
Display detailed information.
.IP --output-mode={parsable,table,json,text}
Display list of block volumes as table, using json format, simple text or using parsable output
(default is table).
.IP --no-truncate
Do not truncate long values during output.
.RE
.TP
oci-network-config show-subnets
Shows subnet information

Columns:
    Name
    OCID
    ipv4 cidr block
    ivp6 cidr block
    VNC Name
    VCN OCID
    Public
    Public ingress
    DNS label
    Domain name
    Lifecycle state
.RS
.IP "-h, --help"
Print a summary of the command line options.
.IP "--ocid SUBNET_NAME"
Show only information of subnet matching name.
.IP "--ocid SUBNET_OCID"
Show  information of subnet matching ocid.
.IP "--details"
Display detailed information.
.IP --output-mode={parsable,table,json,text}
Display list of block volumes as table, using json format, simple text or using parsable output
(default is table).
.IP --no-truncate
Do not truncate long values during output.
.RE
.TP
oci-network-config show-vnics
Shows VNICs information of this instance

Columns:
    Name
    Private IP
    MAC address
    Configuration state
    OCID
    Primary IP
    Subnet name
    Subnet OCID
    Subnet (ipv4) cidr block
    Availability State
    Nic index
    Public IP

For secondary addresses:
    Private IP (ipv4 and ipv6)
    OCID
.RS
.IP "-h, --help"
Print a summary of the command line options.
.IP "--ocid VNIC_NAME"
Show only information of vNIC matching name.
.IP "--ocid VNIC_OCID"
Show  information of vNIC matching ocid.
.IP "--ip-address PRIMARY_IP"
Show information of vNIC matching IP as primary IP.
.IP "--details"
Display detailed information.
.IP --output-mode={parsable,table,json,text}
Display list of block volumes as table, using json format, simple text or using parsable output
(default is table).
.IP --no-truncate
Do not truncate long values during output.
.RE
.TP
oci-network-config show-vnics-all
Shows information on all VNICs of the instance, detailed and not truncated. This is a shortcut for show-vnics --details --no-truncate.
.RE
.TP
oci-network-config configure
Puts in sync network configuration. Add IP configuration for VNICs that are not configured and delete VNICs
that are no longer provisioned.
.RS
.IP "-h, --help"
Print a summary of the command line options.
.IP "-X ITEM, --exclude ITEM"
Persistently exclude
.B ITEM
from automatic configuration/deconfiguration. Use the --include option to include the
.B ITEM
again.
.B ITEM
can be a VNIC OCID, an IP address or a VLAN interface name.
.IP "-I ITEM, --include ITEM"
Include an
.B ITEM
that was previously excluded using the --exclude option in automatic configuration/deconfiguration.
.IP "-n FORMAT, --namespace FORMAT"
When configuring, place interfaces in namespace identified by the given format. Format can
include
.B nic
and
.B vltag
variables. The name defaults to
.B DEF_NS_FORMAT_BM
for
.B BMs
and
.B DEF_NS_FORMAT_VM
for
.B VMs
.
When configuring multiple VNICs ensure the namespaces are unique.
.IP "-r, --sshd"
Start sshd in namespace (if -n or --namespace is present)
.RE
.TP
oci-network-config unconfigure
Deconfigure all VNICs (except the primary).
.RS
.IP "-h, --help"
Print a summary of the command line options.
.IP "-X ITEM, --exclude ITEM"
Persistently exclude
.B ITEM

NEWS
Version 0.11.4
- LINUX-7035 - introduce oci-notify tool to send notification to OCI notification service

Version 0.11.3.3
- Remove python2 support
- LINUX-7986 - fix crash while printing network interfaces on BM shapes with oci-network-config command
- LINUX-7918 - fix regression on -n option of oci-kvm command
- LINUX-7918 - fix free vnic search for oci-kvm command on BM shapes
- LINUX-8011 - interpreter crashes displaying iSCSI information
- LINUX-7672 - fix for python3 byte handling issue.
- support for LVM root filesystem in oci-growfs
- multi vnic support for KVM guests
- removal of libexec/secondary_vnic_all_configure.sh, replaced by python implementation
- oci-kvm, added sanity around parameters passes as part of extra-args option
- add oci-image-migrate code
- move to by-uuid device name for libvirt storage pool build
- LINUX-8008 -  AttributeError: 'OCIVolume' object has no attribute 'HUMAN'
- LINUX-6752 - missing python OCI-SDk dependency

Version 0.10.2
- Update to use Python 3 on OL8

Version 0.10.1
- Added support of libvirt network for KVM guests
- Added support for libvir storage pools

Version 0.9.1
- Added support for Oracle Linux 8 Beta 1

Version 0.9.0
- LINUX-498 - Enhance oci-metadata  for keys to only return key value
              Added suport --value-only
- LINUX-560 - Cleanup utility not preserving permissions/ownerships (fixed)
              same as bug-29260959.
Version 0.8.0

- OLOCITOOLS-10 - added support for updating instance metadata for a specified instance.
                  $ oci-metadata [ -u|--update ] <key>=<value> ...
- OLOCITOOLS-11 - implemented method for retrieving metadata for other compute instances.
                  $ oci-metadata --instance-id <ocid>
- OLOCITOOLS-12 - implemented oci-metadata --export
                  $ oci-metadata --export -g key ..
                - enhanced metadata get by supporting path and wildcard in path.
                  $ oci-metadata -g /path/to/key -g /path/*/key
- OLOCITOOLS-28 - implemented oci-metadata --trim -g   ...
                  trim the path to the last key.
- OLOCITOOLS-24 - removed the metadata cach

- OLOCITOOLS-27 - fixed misleading error message when metadata does not include ssh_authorized_keys.

- fixed bug-28048699 - Needs to handle multiple physical NICs when creating secondary vnics
                       Added option "--nic-index" "to specify which physical NIC the vnic is assigned to.
- fixed bug-28599902 - oci-public-ip returns any public ip it found first.
                       Added an option [-a |--all] to list all public ips for this instance.
- fixed bug-28643343 - fixed most exceptions for oci config error.
- fixed bug-29026433 - oci-iscsi-config don't config secondary ip for second vnic automatically
                       when running delete ip option
- fixed bug-28840370 - local disks  should not count for "max_volumes" when creating iscsi disk
- fixed bug-29026422 - oci-iscsi-config -s " don't list boot partition for VM
- fixed bug-29118662 - oci-network-config detach vnic give wrong message

Version 0.7.1:

- fixed bug 28668447 - ocid needs to allow time for iSCSI connection to recover

Version 0.7:

- bump version for 0.7
- added oci-network-inspector to listing networking information.(OLOCITOOLS-5)
- added oci-growfs to support to grow filesystems (OLOCITOOLS-7)
  Currently only / boot volume is expendable.
- fixed bugs in oci-iscsi-config:
  1. bug-28493320: mis-list sdaa.. as partition of sda. It should be a seperate volume.
  2. bug 28615235 - different behavior for max_volumes setting for 32 or less than 32
  3. bug-28513898 - should not create a volume if it failed to attach
  4. privilege adjustment for create and destroy(root+oci), attach and detach(root)
  5. Bug 27835953 - objects can't be shown due to pagination issue.
  6. Bug 28612818 - A detached volume can't be reattached to the instance saying max limit reached
- fixed bug oci-network-config
  1. bug-28498139 - secondary ips on secondary VNICs not reachable issue.
  2. Bug 28080064 - oci-network-config -s can't list all secondary private ip
  3. Bug 28188853 - oci-network-config report error message when detach vnic
- expanded 'OCI' to 'Oracle Cloud Infrastructure' in man pages, specfiles (OLOCITOOLS-8)

Version 0.6:

- bump version for 0.6
- add support for API access through Instance Principals
- add support for root using a designated user's OCI config files and keys
- oci_utils API automatically detects authentication method to be used
- ocid can discover secondary IP addresses and CHAP user/password using
  OCI API calls, if the Python SDK is configured or Instance Principals used
- network proxy support for making SDK calls
- configuration files for ocid: /etc/oci-utils.d/*
- support configuring the various functions of ocid individually,
  including refresh frequency or turning them off completely.
- ocid saves state and restores all volumes and VNIC configuration after reboot
- oci-network-config: new option: --detach-vnic
- oci-iscsi-config: new option: --destroy-volume
- oci-utils APIs are now thread safe
- NEW tool: oci-image-cleanup - a script that runs a set of cleanup steps to
  prepare the instance for snapshot
- oci-kvm utility rejects attempts to create guests if the required
  virtualization support is not enabled in the image it is being executed on


libexec/oci-network-config.py
#
# Copyright (c) 2017, 2019 Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown
# at http://oss.oracle.com/licenses/upl.

"""
This utility assists with configuring network interfaces on Oracle Cloud
Infrastructure instances.  See the manual page for more information.
"""

import argparse
import logging
import os
import sys

import time
import oci_utils.oci_api
from oci_utils.exceptions import OCISDKError
from oci_utils.vnicutils import VNICUtils

__logger = logging.getLogger("oci-utils.oci-network-config")


def parse_args():
    """
    Parse the command line arguments and return an object representing the
    command line (as returned by argparse's parse_args()).

    Returns
    -------
        The argparse namespace.
    """
    parser = argparse.ArgumentParser(description='Utility for configuring '
                                                 'network interfaces on an '
                                                 'instance running in the '
                                                 'Oracle Cloud '
                                                 'Infrastructure.')
    parser.add_argument('-s', '--show', action='store_true',
                        help='Show information on all provisioning and '
                             'interface configuration. This is the default '
                             'action if no options are given.')
    parser.add_argument('--create-vnic', action='store_true',
                        help='Create a new VNIC and attach it to '
                             'this instance')
    parser.add_argument('--nic-index', action='store', metavar='INDEX',
                        type=int, default=0,
                        help='physical NIC card index. When used with '
                             'the --create-vnic option, assign the new VNIC '
                             'to the specified physical NIC card.')
    parser.add_argument('--detach-vnic', action='store', metavar='VNIC',
                        help='Detach and delete the VNIC with the given OCID'
                             ' or primary IP address')
    parser.add_argument('--add-private-ip', action='store_true',
                        help='Add a secondary private IP to an existing VNIC')
    parser.add_argument('--del-private-ip', action='store', metavar='ADDR',
                        help='delete the secondary private IP address with '
                             'the given IP address')
    parser.add_argument('--private-ip', action='store', metavar='ADDR',
                        help='When used with the --create-vnic or '
                             'add-private-ip options, '
                             'assign the given private IP address to the VNIC')
    parser.add_argument('--subnet', action='store',
                        help='When used with the --create-vnic option, '
                             'connect the new VNIC to the given subnet.')
    parser.add_argument('--vnic-name', action='store', metavar='NAME',
                        help='When used with the --create-vnic option, '
                             'use NAME as the display name of the new VNIC')
    parser.add_argument('--assign-public-ip', action='store_true',
                        help='When used with the --create-vnic option, '
                             'assign a public IP address to the new VNIC.')
    parser.add_argument('--vnic', action='store', metavar='OCID',
                        help='When used with the --add-private-ip option, '
                             'assign the private IP to the given VNIC')
    parser.add_argument('-a', '--auto', '-c', '--configure',
                        action='store_true',
                        help='Add IP configuration for VNICs that are not '
                             'configured and delete for VNICs that are no '
                             'longer provisioned.')
    parser.add_argument('-d', '--deconfigure', action='store_true',
                        help='Deconfigure all VNICs (except the primary). If '
                             'a -e option is also present only the secondary '
                             'IP address(es) are deconfigured.')
    parser.add_argument('-e', nargs=2, metavar=('IP_ADDR', 'VNIC_OCID'),
                        dest='sec_ip', action='append',
                        help='Secondary private IP address to configure or '
                             'deconfigure.  Use in conjunction with -c or -d.')
    parser.add_argument('-n', '--ns', action='store', metavar='FORMAT',
                        help='When configuring, place interfaces in namespace '
                             'identified by the given format. Format can '
                             'include $nic and $vltag variables.')
    parser.add_argument('-r', '--sshd', action='store_true',
                        help='Start sshd in namespace (if -n is present)')
    parser.add_argument('-X', '--exclude', metavar='ITEM', action='append',
                        type=str, dest='exclude',
                        help='Persistently exclude ITEM from automatic '
                             'configuration/deconfiguration.  Use the '
                             '--include option to include the ITEM again.')
    parser.add_argument('-I', '--include', metavar='ITEM', action='append',
                        type=str, dest='include',
                        help='Include an ITEM that was previously excluded '
                             'using the --exclude option in automatic '
                             'configuration/deconfiguration.')
    parser.add_argument('--quiet', '-q', action='store_true',
                        help='Suppress information messages')
    args = parser.parse_args()
    return args


def get_oci_api_session(opt_name=None):
    """
    Ensure the OCI SDK is available if the option is not None.

    Parameters
    ----------
    opt_name : str
        Operation name currently been exceuted (used for logging).

    Returns
    -------
        OCISession
            The session or None if cannot get one
    """
    sess = None
    if oci_utils.oci_api.HAVE_OCI_SDK:
        try:
            sess = oci_utils.oci_api.OCISession()
        except Exception as e:
            sdk_error = str(e)
            if opt_name is not None:
                __logger.error("To use the %s option, you need to "
                               "install and configure the OCI Python SDK "
                               "(python-oci-sdk)\n" % opt_name)
                __logger.error(sdk_error)
            else:
                __logger.error("Failed to access OCI services: %s" % sdk_error)

    return sess


def api_show_network_config():
    """
    Show the current network configuration of the instance based on
    information obtained through OCI API calls, if the OCI SDK is
    configured.

    Returns
    -------
       No return value.
    """

    sess = get_oci_api_session()
    if sess is None:
        __logger.error("Failed to get API session.")
        return
    inst = sess.this_instance()
    if inst is None:
        __logger.error("Failed to get information from OCI.")
        return
    vnics = inst.all_vnics()
    i = 1
    print "VNIC configuration for instance %s" % inst.get_display_name()
    print
    for vnic in vnics:
        primary = ""
        if vnic.is_primary():
            primary = " (primary)"
        print "VNIC %d%s: %s" % (i, primary, vnic.get_display_name())
        print "     Hostname: %s" % vnic.get_hostname()
        print "     OCID: %s" % vnic.get_ocid()
        print "     MAC address: %s" % vnic.get_mac_address()
        print "     Public IP address: %s" % vnic.get_public_ip()
        print "     Private IP address: %s" % vnic.get_private_ip()

        _subn = vnic.get_subnet()
        if _subn is not None:
            print "     Subnet: %s (%s)" % (_subn.get_display_name(), _subn)
        else:
            print "     Subnet: Not found"

        privips = vnic.all_private_ips()
        if len(privips) > 0:
            print "     Private IP addresses:"
            for privip in privips:
                print "         IP address: %s" % privip.get_address()
                print "         OCID: %s" % privip.get_ocid()
                print "         Hostname: %s" % privip.get_hostname()
                print "         Subnet: %s (%s)" % \
                    (privip.get_subnet().get_display_name(),
                     privip.get_subnet().get_cidr_block())
                print
        else:
            print
        i += 1


def do_show_network_config(vnic_utils):
    """
    Display the currect network interface configuration as well as the
    VNIC configuration from OCI.

    Parameters
    ----------
    vnic_utils :
        The VNIC configuration instance.

    Returns
    -------
        No return value.
    """
    if oci_utils.oci_api.HAVE_OCI_SDK:
        api_show_network_config()

    __logger.info("Operating System level network configuration")

    (ret, out) = vnic_utils.get_network_config()
    if ret:
        __logger.error("Failed to execute the VNIC configuration script.")
    else:
        print "%s" % out


def do_detach_vnic(detach_options, vnic_utils):
    """
    Detach and delete the VNIC with the given ocid or primary ip address

    Parameters
    ----------
    detach_options : namespace
        The argparse namespace.
    vnic_utils :
        The VNIC configuration instance.

    Returns
    -------
        No return value on success;.

    Raises
    ------
        StandardError
            if session cannot be acquired
            if the VNIC cannot be detached

    """
    # needs the OCI SDK installed and configured
    sess = get_oci_api_session(opt_name="--detach-vnic")
    if sess is None:
        raise StandardError("Failed to get API session.")
    vnics = sess.this_instance().all_vnics()
    for vnic in vnics:
        if vnic.get_ocid() == detach_options.detach_vnic or \
           vnic.get_private_ip() == detach_options.detach_vnic:
            if not vnic.is_primary():
                vnic_utils.delete_all_private_ips(vnic.get_ocid())
                vnic.detach()
                break
            else:
                raise StandardError("The primary VNIC cannot be detached.")
    return sess.this_shape()


def do_create_vnic(create_options):
    """
    Create and attach a VNIC to this instance.

    Parameters
    ----------
    create_options :
        The VNIC configuration instance.

    Returns
    -------
        No return value on success; errors out with return value 1 otherwise.

    Raises
    ------
        StandardError
            if session cannot be acquired
    """
    # needs the OCI SDK installed and configured
    sess = get_oci_api_session(opt_name="--create-vnic")
    if sess is None:
        raise StandardError("Failed to get API session.")
    subnet_id = None
    if create_options.subnet:
        if create_options.subnet.startswith('ocid1.subnet.'):
            subnet = sess.get_subnet(create_options.subnet)
            if subnet is None:
                raise StandardError(
                    "Subnet not found: %s\n" % create_options.subnet)
            else:
                subnet_id = subnet.get_ocid()
        else:
            subnets = sess.find_subnets(create_options.subnet)
            if len(subnets) == 0:
                raise StandardError(
                    "No subnet matching %s found\n" % create_options.subnet)
            elif len(subnets) > 1:
                __logger.error("More than one subnet matching %s found:\n"
                               % create_options.subnet)
                for sn in subnets:
                    __logger.error("   %s\n" % sn.get_display_name())
                raise StandardError("More than one subnet matching")
            subnet_id = subnets[0].get_ocid()
    try:
        vnic = sess.this_instance().attach_vnic(
            private_ip=create_options.private_ip,
            assign_public_ip=create_options.assign_public_ip,
            subnet_id=subnet_id,
            nic_index=create_options.nic_index,
            display_name=create_options.vnic_name)
    except OCISDKError as e:
        raise StandardError('Failed to create VNIC: %s' % e)

    public_ip = vnic.get_public_ip()
    if public_ip is not None:
        __logger.info(
            'creating VNIC: %s (public IP %s)' % (vnic.get_private_ip(),
                                                  public_ip))
    else:
        __logger.info('creating VNIC: %s' % vnic.get_private_ip())


def do_add_private_ip(vnic_utils, add_options):
    """
    Add a secondary private IP for an existing VNIC.

    Parameters
    ----------
    vnic_utils : VNICUtils
        The VNICUtils helper instance.
    add_options : namespace
        The argparse namespace.

    Returns
    -------
        tuple
            (private_IP,vnic_ocid) for the new IP on success; errors out with
            return value 1 otherwise.

    Raises
    ------
        StandardError
            On any error.
    """
    # needs the OCI SDK installed and configured
    sess = get_oci_api_session(opt_name="--add-private-ip")
    if sess is None:
        raise StandardError("Failed to get API session.")

    if add_options.vnic:
        if add_options.vnic.startswith('ocid1.vnic.'):
            vnic = sess.get_vnic(add_options.vnic)
            if vnic is None:
                raise StandardError("VNIC not found: %s" % add_options.vnic)
            else:
                pass
        else:
            raise StandardError("Invalid VNIC OCID: %s" % add_options.vnic)

    else:
        vnics = sess.this_instance().all_vnics()
        if len(vnics) > 1:
            __logger.error("More than one VNIC found."
                           "Use the --vnic option to select the one to add "
                           "a secondary IP for:")
            for vnic in vnics:
                __logger.error("   %s: %s" % (vnic.get_private_ip(),
                                              vnic.get_ocid()))
            raise StandardError("Too many VNICs found")
        vnic = vnics[0]
    try:
        priv_ip = vnic.add_private_ip(private_ip=add_options.private_ip)
    except OCISDKError as e:
        raise StandardError('Failed to provision private IP: %s' % e)

    __logger.info(
        'provisioning secondary private IP: %s' % priv_ip.get_address())
    vnic_utils.add_private_ip(priv_ip.get_address(), vnic.get_ocid())
    return priv_ip.get_address(), vnic.get_ocid()


def do_del_private_ip(vnic_utils, delete_options):
    """
    Delete a secondary private IP

    Parameters
    ----------
    vnic_utils :
        The VNIC configuration instance.
    delete_options :
        The argparse namespace.

    Returns
    -------
        No return value on success; errors out with return value 1 otherwise.

    Raises
    ------
    StandardError
        error getting session
    """
    # needs the OCI SDK installed and configured
    sess = get_oci_api_session(opt_name="--del-private-ip")
    if sess is None:
        raise StandardError("Failed to get API session.")
    # find the private IP
    priv_ip = sess.this_instance().find_private_ip(
        delete_options.del_private_ip)
    if priv_ip is None:
        raise StandardError(
            "Secondary private IP not found: %s" %
            delete_options.del_private_ip)

    if priv_ip.is_primary():
        raise StandardError("Cannot delete IP %s, it is the primary private "
                            "address of the VNIC." %
                            delete_options.del_private_ip)
    vnic_id = None
    try:
        vnic_id = priv_ip.get_vnic_ocid()
    except Exception:
        pass

    if not priv_ip.delete():
        raise StandardError('failed to delete secondary private IP %s' %
                            delete_options.del_private_ip)

    __logger.info('deconfigure secondary private IP %s' %
                  delete_options.del_private_ip)
    # delete from vnic_info and de-configure the interface
    return vnic_utils.del_private_ip(delete_options.del_private_ip, vnic_id)


def main():
    """
    Main

    Returns
    -------
        int
            0 on success;
            1 on failure.
    """
    args = parse_args()

    if os.geteuid() != 0:
        __logger.error("You must run this program with root privileges")
        return 1

    if args.create_vnic:
        if args.add_private_ip:
            __logger.error(
                "Cannot use --create-vnic and --add-private-ip at the "
                "same time")
            return 1
        try:
            do_create_vnic(args)
        except StandardError as e:
            __logger.debug('cannot create the VNIC', exc_info=True)
            __logger.error('cannot create the VNIC: %s' % str(e))
            return 1
    try:
        vnic_utils = VNICUtils()
        vnic_info = vnic_utils.get_vnic_info()[1]
    except Exception as e:
        __logger.warning("OCI SDK Error: %s" % str(e))
        __logger.exception('OCI SDK Error')
        return 1

    shape = None
    if args.detach_vnic:
        try:
            shape = do_detach_vnic(args, vnic_utils)
            time.sleep(10)
        except StandardError as e:
            __logger.error(str(e))
            return 1

    if args.ns:
        vnic_utils.set_namespace(args.ns)

    if args.sshd:
        vnic_utils.set_sshd(args.sshd)

    excludes = vnic_info['exclude']
    if excludes is None:
        excludes = []

    ret = 0
    out = ""
    if args.add_private_ip:
        try:
            (ip, vnic_id) = do_add_private_ip(vnic_utils, args)
            __logger.info("IP %s has been assigned to vnic %s." % (ip, vnic_id))
        except StandardError as e:
            __logger.error('failed ot add private ip: %s' % str(e))
            return 1

    elif args.del_private_ip:
        try:
            (ret, out) = do_del_private_ip(vnic_utils, args)
        except StandardError as e:
            __logger.error('failed ot delete private ip: %s' % str(e))
            return 1

    if args.exclude:
        for exc in args.exclude:
            if args.include and exc in args.include:
                __logger.error(
                    "Invalid arguments: item both included and excluded: %s"
                    % exc)
            vnic_utils.exclude(exc)
        excludes = vnic_info['exclude']
    if args.include:
        for inc in args.include:
            vnic_utils.include(inc)
        excludes = vnic_info['exclude']

    if excludes and not args.quiet:
        if __logger.isEnabledFor(logging.INFO):
            __logger.info(
                "Info: Addresses excluded from automatic configuration: %s" %
                ", ".join(excludes))

    if args.auto or args.create_vnic or args.add_private_ip:
        (ret, out) = vnic_utils.auto_config(quiet=args.quiet, show=args.show,
                                            sec_ip=args.sec_ip)
    elif args.detach_vnic and shape and shape.startswith("BM"):
        (ret, out) = vnic_utils.auto_config(quiet=args.quiet, show=args.show,
                                            sec_ip=args.sec_ip)
    elif args.deconfigure:
        (ret, out) = vnic_utils.auto_deconfig(quiet=args.quiet, show=args.show,
                                              sec_ip=args.sec_ip)

    if ret:
        __logger.error("Failed to execute the VNIC configuration script.")
    if out:
        __logger.debug(str(out))

    if not args.quiet or args.show:
        do_show_network_config(vnic_utils)

    return 0


if __name__ == "__main__":
    sys.exit(main())
